starknet-providers/src/jsonrpc/mod.rs

```diff
@@ -388,53 +393,56 @@ where
     /// Submit a new transaction to be added to the chain
     pub async fn add_invoke_transaction(
         &self,
-        function_invocation: &FunctionCall,
-        signature: Vec<FieldElement>,
-        max_fee: FieldElement,
-        version: FieldElement,
+        invoke_transaction: &BroadcastedInvokeTransaction,
     ) -> Result<InvokeTransactionResult, JsonRpcClientError<T::Error>> {
         self.send_request(
             JsonRpcMethod::AddInvokeTransaction,
-            [
-                serde_json::to_value(function_invocation)?,
-                serde_json::to_value(FeltArray(signature))?,
-                serde_json::to_value(Felt(max_fee))?,
-                serde_json::to_value(Felt(version))?,
-            ],
+            [serde_json::to_value(BroadcastedTransaction::Invoke(
+                invoke_transaction.clone(),
+            ))?],
         )
         .await
     }
```

Comment:
Asking the user to pass in reference only to `clone()` inside is bad practice. The user might have an object he's willing to give up ownership for, but this function here is forcing an extra clone, which could be expensive.

The responsibility for cloning is on the user, _if_ we have to take ownership.

Reply:
Actually, there's no need to take ownership here at all. You shouldn't need to wrap the object in a `BroadcastedTransaction` enum in the first place. It's just for serialization after all.

Reply:
Moved tagging with type from the enum to structs as it is only used for serialization